### ISSUE TITLE
'callback constraint is not a valid callable ' in the upload 

### ISSUE THREAD

**Issue by aminealaoui** [sonata-project/SonataMediaBundle#1101]
#### Sonata packages

``` bash
$ composer show sonata-project/*
sonata-project/admin-bundle              3.4.0           The missing Symfony Admin Generator
sonata-project/block-bundle              3.1.1           Symfony SonataBlockBundle
sonata-project/cache                     1.0.7           Cache library
sonata-project/classification-bundle     3.x-dev ad97ae7 Symfony SonataClassificationBundle
sonata-project/core-bundle               3.0.3           Symfony SonataCoreBundle
sonata-project/datagrid-bundle           2.2             Symfony SonataDatagridBundle
sonata-project/doctrine-extensions       1.0.2           Doctrine2 behavioral extensions
sonata-project/doctrine-orm-admin-bundle 3.0.5           Symfony Sonata / Integrate Doctrine ORM into the SonataAdminBundle
sonata-project/easy-extends-bundle       2.1.10          Symfony SonataEasyExtendsBundle
sonata-project/exporter                  1.6.0           Lightweight Exporter library
sonata-project/intl-bundle               2.2.4           Symfony SonataIntlBundle
sonata-project/media-bundle              3.1.0           Symfony SonataMediaBundle
sonata-project/notification-bundle       3.0.0           Symfony SonataNotificationBundle

```
#### Symfony packages

``` bash
$ composer show symfony/*
symfony/monolog-bundle     2.11.1  Symfony MonologBundle
symfony/phpunit-bridge     v3.1.3  Symfony PHPUnit Bridge
symfony/polyfill-apcu      v1.2.0  Symfony polyfill backporting apcu_* functions to lower PHP versions
symfony/polyfill-intl-icu  v1.2.0  Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-mbstring  v1.2.0  Symfony polyfill for the Mbstring extension
symfony/polyfill-php56     v1.2.0  Symfony polyfill backporting some PHP 5.6+ features to lower PHP versions
symfony/polyfill-php70     v1.2.0  Symfony polyfill backporting some PHP 7.0+ features to lower PHP versions
symfony/polyfill-util      v1.2.0  Symfony utilities for portability of PHP codes
symfony/security-acl       v3.0.0  Symfony Security Component - ACL (Access Control List)
symfony/swiftmailer-bundle v2.3.11 Symfony SwiftmailerBundle
symfony/symfony            v3.1.3  The Symfony PHP framework

```
#### PHP version

``` bash
$ php -v
PHP 5.6.15 (cli) (built: Oct 29 2015 12:40:36)
Copyright (c) 1997-2015 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2015 Zend Technologies

```
## Subject

Hello, i'm trying to install sonata media bundle (with classification) with symfony 3, 
i get the interface for uploading the pictures, but when i click in the save button i have this error :

``` bash
CRITICAL - Uncaught PHP Exception Symfony\Component\Validator\Exception\ConstraintDefinitionException: "{"methods":["isStatusErroneous"]} targeted by Callback constraint is not a valid callable" at C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\Validator\Constraints\CallbackValidator.php line 43 
```

any suggestion please

**Comment by greg0ire** [comment]
> any suggestion please 

Fill the issue template.

**Comment by aminealaoui** [comment]
like this @greg0ire ?

**Comment by greg0ire** [comment]
Better! Can you also provide a stack trace? It should help a lot!

**Comment by aminealaoui** [comment]
```bash
CRITICAL - Uncaught PHP Exception Symfony\Component\Validator\Exception\ConstraintDefinitionException: "{"methods":["isStatusErroneous"]} targeted by Callback constraint is not a valid callable" at C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\Validator\Constraints\CallbackValidator.php line 43 
```
```bash
Symfony\Component\Validator\Exception\ConstraintDefinitionException: {"methods":["isStatusErroneous"]} targeted by Callback constraint is not a valid callable
    at n/a
        in C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\Validator\Constraints\CallbackValidator.php line 43

    at Symfony\Component\Validator\Constraints\CallbackValidator->validate(object(Media), object(Callback))
        in C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\Validator\Validator\RecursiveContextualValidator.php line 841

    at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateInGroup(object(Media), '0000000034cc4fd0000000004b969d20', object(ClassMetadata), 'Default', object(ExecutionContext))
        in C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\Validator\Validator\RecursiveContextualValidator.php line 552

    at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateClassNode(object(Media), '0000000034cc4fd0000000004b969d20', object(ClassMetadata), 'data', array('Default'), null, '1', object(ExecutionContext))
        in C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\Validator\Validator\RecursiveContextualValidator.php line 358

    at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateObject(object(Media), 'data', array('Default'), '1', object(ExecutionContext))
        in C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\Validator\Validator\RecursiveContextualValidator.php line 153

    at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validate(object(Media), null, 'Default')
        in C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\Form\Extension\Validator\Constraints\FormValidator.php line 51

    at Symfony\Component\Form\Extension\Validator\Constraints\FormValidator->validate(object(Form), object(Form))
        in C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\Validator\Validator\RecursiveContextualValidator.php line 841

    at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateInGroup(object(Form), '0000000034cc4f74000000004b969d20', object(ClassMetadata), 'Default', object(ExecutionContext))
        in C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\Validator\Validator\RecursiveContextualValidator.php line 552

    at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateClassNode(object(Form), '0000000034cc4f74000000004b969d20', object(ClassMetadata), '', array('Default'), null, '1', object(ExecutionContext))
        in C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\Validator\Validator\RecursiveContextualValidator.php line 358

    at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateObject(object(Form), '', array('Default'), '1', object(ExecutionContext))
        in C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\Validator\Validator\RecursiveContextualValidator.php line 153

    at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validate(object(Form), null, null)
        in C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\Validator\Validator\RecursiveValidator.php line 115

    at Symfony\Component\Validator\Validator\RecursiveValidator->validate(object(Form))
        in C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\Form\Extension\Validator\EventListener\ValidationListener.php line 55

    at Symfony\Component\Form\Extension\Validator\EventListener\ValidationListener->validateForm(object(FormEvent), 'form.post_bind', object(EventDispatcher))
        in  line 

    at call_user_func(array(object(ValidationListener), 'validateForm'), object(FormEvent), 'form.post_bind', object(EventDispatcher))
        in C:\wamp\www\monprojet\var\cache\dev\classes.php line 2144

    at Symfony\Component\EventDispatcher\EventDispatcher->doDispatch(array(array(object(ValidationListener), 'validateForm'), array(object(DataCollectorListener), 'postSubmit')), 'form.post_bind', object(FormEvent))
        in C:\wamp\www\monprojet\var\cache\dev\classes.php line 2059

    at Symfony\Component\EventDispatcher\EventDispatcher->dispatch('form.post_bind', object(FormEvent))
        in C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\EventDispatcher\ImmutableEventDispatcher.php line 43

    at Symfony\Component\EventDispatcher\ImmutableEventDispatcher->dispatch('form.post_bind', object(FormEvent))
        in C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\Form\Form.php line 655

    at Symfony\Component\Form\Form->submit(array('providerName' => 'sonata.media.provider.image', 'category' => '1', '_token' => 'NawzzjMYAcVWjljIqwPC7WVvHwFfJ3rX2Hdzhs6GLHw', 'binaryContent' => object(UploadedFile)), true)
        in C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\Form\Extension\HttpFoundation\HttpFoundationRequestHandler.php line 116

    at Symfony\Component\Form\Extension\HttpFoundation\HttpFoundationRequestHandler->handleRequest(object(Form), object(Request))
        in C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\Form\Form.php line 488

    at Symfony\Component\Form\Form->handleRequest(object(Request))
        in C:\wamp\www\monprojet\vendor\sonata-project\admin-bundle\Controller\CRUDController.php line 501

    at Sonata\AdminBundle\Controller\CRUDController->createAction()
        in C:\wamp\www\monprojet\vendor\sonata-project\media-bundle\Controller\MediaAdminController.php line 40

    at Sonata\MediaBundle\Controller\MediaAdminController->createAction(object(Request))
        in  line 

    at call_user_func_array(array(object(MediaAdminController), 'createAction'), array(object(Request)))
        in C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\HttpKernel\HttpKernel.php line 153

    at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), '1')
        in C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\HttpKernel\HttpKernel.php line 68

    at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), '1', true)
        in C:\wamp\www\monprojet\vendor\symfony\symfony\src\Symfony\Component\HttpKernel\Kernel.php line 169

    at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
        in C:\wamp\www\monprojet\web\app_dev.php line 30
```

**Comment by greg0ire** [comment]
Oh wow… this constraint looks nothing like the docs. Can you try replacing [this options tag](https://github.com/sonata-project/SonataMediaBundle/blob/3.x/Resources/config/validation.xml#L5) with just the name of the method?

**Comment by aminealaoui** [comment]
thanks @greg0ire  the code work whene i change this part
```bash
        <constraint name="Callback">
            <option name="methods">
                <value>isStatusErroneous</value>
            </option>
        </constraint>
```
with 

```bash
       <constraint name="Callback">isStatusErroneous</constraint>
```

but now i get this error 

```bash
Impossible to access a key ("default") on a boolean variable ("") in SonataAdminBundle:CRUD:base_edit_form.html.twig at line 58
```

**Comment by greg0ire** [comment]
Can you link to the code of the template on GitHub?

**Comment by aminealaoui** [comment]
https://gist.github.com/aminealaoui/387bc00e2c559499b2ce5088db7dfb92#file-base_edit_form-html-twig-L58

**Comment by greg0ire** [comment]
What I meant was that : https://github.com/sonata-project/SonataAdminBundle/blob/3.x/Resources/views/CRUD/base_edit_form.html.twig#L58

**Comment by aminealaoui** [comment]
this is the calls of setFormTabs
```bash
    protected function setTabs(array $tabs)
    {
        $this->admin->setFormTabs($tabs);
    }
```
and this is the setFormTabs 
```bash
    public function setFormTabs(array $formTabs)
    {
        $this->formTabs = $formTabs;
    }
```

**Comment by greg0ire** [comment]
These are method definitions, what you I asked for is method calls… but then it's not possible that these methods were called with a boolean, otherwise there would have been an error because of the type hinting. Maybe there are other ways to assign `$this->formTabs` ?

**Comment by greg0ire** [comment]
Ok, so I looked, and the default value for `$formTabs` is false, this is why you get an error. Because there are no calls to `setFormTabs`.

**Comment by greg0ire** [comment]
Try using `with()` to define a group, it should fix your problem.

**Comment by aminealaoui** [comment]
please can you explain more "using with() to define a group"

thanks

**Comment by greg0ire** [comment]
When building the form, make sure you use `$formMapper->with()`. I'm on my phone, difficult to explain more

**Comment by vub** [comment]
I had a same issues.
I change:
` <constraint name="Callback">isStatusErroneous</constraint>`
and also
add **with('something')** in **configureFormFields** function on **Mediabundle\Admin\BaseMediaAdmin.php**
but it still not work.
It's just create a tab name "something" in create media screen.

**Comment by binhle410** [comment]
Same here, checked out master and used the default MediaAdmin object.
The culprit is that src\Sonata\MediaBundle\Admin\BaseMediaAdmin::getNewInstance
cannot read providerName and therefore cannot create the new Media which leads to the error.
It cannot read providerName due to
```
$media->setProviderName($this->getRequest()->get(sprintf('%s[providerName]', $this->getUniqid()), null, true));
```
this line of code will set the value NULL to providerName of media object because $this->getRequest()->get(....[providerName[) cannot find the posted data which was **s57d7a9c90fcff[providerName]:sonata.media.provider.youtube**
This data was perceived as a 2 dimension array and hence the array_key_exists method call of Symfony would return NULL because it couldn't find the posted data.

**Comment by greg0ire** [comment]
There are several problems in this issues. Can anyone make PRs for the solutions that were found?

**Comment by idchlife** [comment]
Any info on resolving?

**Comment by mlivan** [comment]
I have the same problem... Any info?

**Comment by greg0ire** [comment]
@mlivan see #1148 and #1115 @elmpp might be trying to carry #1115 . If you want this fixed, please work together to provide a BC version of this PRs.

**Comment by jordisala1991** [comment]
All this should be fixed on SonataMediaBundle 3.4.0.
